service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/dataproccluster/UpdateDataprocClusterStep.java

```diff
@@ -0,0 +1,149 @@
+import java.util.ArrayList;
+import java.util.List;
+import org.springframework.http.HttpStatus;
+
+public class UpdateDataprocClusterStep implements Step {
+
+  private final ControlledDataprocClusterResource resource;
+  private final CrlService crlService;
+
+  // Mask paths for cluster config fields that are updatable
+  private static final String NUM_WORKERS = "config.worker_config.num_instances";
+  private static final String NUM_SECONDARY_WORKERS =
+      "config.secondary_worker_config.num_instances";
+  private static final String AUTOSCALING_POLICY = "config.autoscaling_config.policy_uri";
+  // Only clusters created with scheduled deletion enabled can be updated. Removing scheduled
+  // deletion from a cluster, i.e. settling lifecycleConfig to be empty is irreversible.
+  private static final String IDLE_DELETE_TTL = "config.lifecycle_config.idle_delete_ttl";
+  private static final String AUTO_DELETE_TTL = "config.lifecycle_config.auto_delete_ttl";
+  private static final String AUTO_DELETE_TIME = "config.lifecycle_config.auto_delete_time";
+
+  public UpdateDataprocClusterStep(
+      ControlledDataprocClusterResource resource, CrlService crlService) {
+    this.resource = resource;
+    this.crlService = crlService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    ApiControlledDataprocClusterUpdateParameters updateParameters =
+        flightContext
+            .getInputParameters()
+            .get(UPDATE_PARAMETERS, ApiControlledDataprocClusterUpdateParameters.class);
+    if (updateParameters == null) {
+      return StepResult.getStepResultSuccess();
+    }
+    return updateCluster(updateParameters);
+  }
+
+  // Restore the previous values of the update parameters
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    FlightMap workingMap = flightContext.getWorkingMap();
+    ApiControlledDataprocClusterUpdateParameters prevParameters =
+        FlightUtils.getRequired(
+            workingMap,
+            PREVIOUS_UPDATE_PARAMETERS,
+            ApiControlledDataprocClusterUpdateParameters.class);
+    return updateCluster(prevParameters);
+  }
+
+  // Update the cluster with the given update parameters.
+  private StepResult updateCluster(ApiControlledDataprocClusterUpdateParameters updateParameters) {
+    ClusterName clusterName = resource.toClusterName();
+    List<String> updateMaskPaths = new ArrayList<>();
+    Cluster updateClusterMask =
+        new Cluster().setClusterName(clusterName.name()).setConfig(new ClusterConfig());
+
+    // Apply the update parameters to the cluster config mask
+    if (updateParameters.getNumPrimaryWorkers() != null) {
+      updateMaskPaths.add(NUM_WORKERS);
+      updateClusterMask
+          .getConfig()
+          .setWorkerConfig(
+              new InstanceGroupConfig().setNumInstances(updateParameters.getNumPrimaryWorkers()));
```

Comment:
Are there any other values that could be set on the `InstanceGroupConfig` object? If so, this needs to be something like 
```
existingConfig = updateClusterMask.getConfig().getWorkerConfig();
updateClusterMask.getConfig().setWorkerConfig(existingConfig().setNumInstances(...)))
...
```
otherwise, you'll lose all the other values from this config. Same for other configs below

Reply:
The docs outlines a small set of updatable parameters. For `InstanceGroupConfig`, its just `numInstances`. The dataproc patch api is a bit confusing, perhaps they chose to reuse the `Cluster` model to hold the updatable fields as opposed to a flat list of parameters.